### PR TITLE
Add separate export path for frontend-specific features

### PR DESCRIPTION
**Pull request by paulomarg**
### WHY are these changes introduced?

We'll want to include some frontend-specific features in this package in the future, and currently the code structure is centered around server-side-only code, which makes it difficult to find where to put that new code as we add it.

### WHAT is this pull request doing?

This PR is simply moving things around into separate `server` and `react` folders and adding an ESLint rule to ensure we're not cross-importing files

## Type of change

- [x] Minor: New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have used `yarn changeset` to create a draft changelog entry (do NOT update the `CHANGELOG.md` files manually)
- [x] I have added/updated tests for this change
- [x] I have documented new APIs/updated the documentation for modified APIs (for public APIs)

### CODE REVIEW

File: .changeset/lovely-swans-repeat.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+'@shopify/shopify-app-remix': minor
+---
+
+- Internally rearranged source files to create a better separation between backend and frontend code, so we can add frontend-specific exports in the future.
+- A new export path `@shopify/shopify-app-remix/react` will now contain those frontend exports.
```
Comment (byrichardpowell) [inline .changeset/lovely-swans-repeat.md:6]: Would only ship this change once we have something to put in the react export?  I'm in two minds here.  Shipping often vs shipping a large changelog that doesn't contain much the partner needs to action.

File: .gitignore [hunk 2]
```diff
@@ -7,10 +7,7 @@ coverage/
 tsdocs/
 tsconfig.tsbuildinfo
 database.sqlite
-packages/**/*.d.ts
-packages/**/*.d.ts.map
-packages/**/*.js
-packages/**/*.js.map
+packages/**/build
```
Comment (byrichardpowell) [inline .gitignore:10]: For my own understanding, why this change?

File: packages/shopify-app-remix/package.json [hunk 3]
```diff
@@ -12,15 +12,18 @@
...
   "exports": {
     ".": {
-      "import": "./build/cjs/index.js",
-      "require": "./build/cjs/index.js",
-      "types": "./build/ts/index.d.ts"
+      "types": "./build/ts/server/index.d.ts",
+      "default": "./build/cjs/server/index.js"
     },
-    "./adapters/*": "./build/cjs/adapters/*/index.js"
+    "./adapters/*": "./build/cjs/adapters/*/index.js",
```
Comment (byrichardpowell) [inline packages/shopify-app-remix/package.json:22]: Should adaptors be provided by the server export and aliased?

File: packages/shopify-app-remix/src/__tests__/test-helper.ts [hunk 4]
```diff
@@ -33,203 +32,3 @@ export function testConfig(
...
-    expect(searchParams.get('host')).toBe(host);
-  }
-}
-
-export function expectLoginRedirect(response: Response) {
-  expect(response.status).toBe(302);
-
-  const {pathname} = new URL(response.headers.get('location')!, APP_URL);
-  expect(pathname).toBe('/auth/login');
-}
```
Comment (byrichardpowell) [inline packages/shopify-app-remix/src/__tests__/test-helper.ts:235]: Is the thinking here that `/src/__tests__/test-helper.ts` is for shared test helpers whereas `/src/server/__tests__test-helper` is for server specific stuff? If so, I like it the thinking... I'm just wondering when `testConfig` would be used by anythign other than server specific stuff?

The React stuff we talked about doesn't need the test config for example

File: packages/shopify-app-remix/src/react/.eslintrc.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+module.exports = {
+  rules: {
+    'no-restricted-imports': [
+      'error',
+      {
+        patterns: [
+          {
+            group: ['../server/*'],
+            message:
+              'Importing from src/server in src/react is not allowed as they must be kept separate',
```
Comment (byrichardpowell) [inline packages/shopify-app-remix/src/react/.eslintrc.js:10]: Worthwhile explaining why? E.g: It will blow up Remix due to different runtimes.

File: packages/shopify-app-remix/src/server/.eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+module.exports = {
+  rules: {
+    'no-restricted-imports': [
+      'error',
+      {
+        patterns: [
+          {
+            group: ['../react/*'],
+            message:
+              'Importing from src/react in src/server is not allowed as they must be kept separate',
```
Comment (byrichardpowell) [inline packages/shopify-app-remix/src/server/.eslintrc.js:10]: Worthwhile explaining why? E.g: It will blow up Remix due to different runtimes.

File: tsconfig.base.json [hunk 7]
```diff
@@ -16,6 +16,7 @@
     "noUnusedLocals": true,
     "importHelpers": true,
     "strictNullChecks": true,
+    "stripInternal": true,
```
Comment (byrichardpowell) [inline tsconfig.base.json:19]: Why this change? I missed anything in this PR that needs this.

**Review by byrichardpowell — COMMENTED** [review]
Much prefer this organisation.

My only other question is on the benefits of something like

```
import {...} from "@shopify/shopify-app-remix/react
import {...} from "@shopify/shopify-app-remix/server
import "@shopify/shopify-app-remix/server/adaptors/node
```

Then we always make it explicit if it's a server or react export.

File: .gitignore (hunk 2 above)
Comment (paulomarg) [reply .gitignore:10]: It's just fewer things to keep in mind - everything we build goes into `build`, so we can just blanket ignore the folder as a whole, which allows us to have js files elsewhere if we ever need them.

File: .changeset/lovely-swans-repeat.md (hunk 1 above)
Comment (paulomarg) [reply .changeset/lovely-swans-repeat.md:6]: I have a different PR coming with the `AppProvider` that will be exported out of this path. We can fix up the changelog once everything is combined to ensure it's reasonable.

File: packages/shopify-app-remix/package.json (hunk 3 above)
Comment (paulomarg) [reply packages/shopify-app-remix/package.json:22]: I'd rather keep them fully separate because that makes it harder to accidentally import them and break apps.

File: packages/shopify-app-remix/src/__tests__/test-helper.ts (hunk 4 above)
Comment (paulomarg) [reply packages/shopify-app-remix/src/__tests__/test-helper.ts:235]: Yes, that was the thinking, but that's a great point about the config - we shouldn't ever need that for react stuff. I'll move that over and delete the file if it's empty.

File: tsconfig.base.json (hunk 7 above)
Comment (paulomarg) [reply tsconfig.base.json:19]: We have one field somewhere that's marked as `@internal`, this removes it from the exported package types. I think that comes with the AppProvider component though, I broke the PR up in two and this landed in the wrong place. We'll still need it though.

**Comment by paulomarg** [comment]
> Much prefer this organisation.
> 
> My only other question is on the benefits of something like
> 
> ```
> import {...} from "@shopify/shopify-app-remix/react
> import {...} from "@shopify/shopify-app-remix/server
> import "@shopify/shopify-app-remix/server/adaptors/node
> ```
> 
> Then we always make it explicit if it's a server or react export.

No objections, but we will still have to alias `server` to `.` so it may be more confusing than helpful.

**Comment by byrichardpowell** [comment]
> No objections, but we will still have to alias server to . so it may be more confusing than helpful.

Yeah, good point on the aliasing.  Long term, I think aliasing allows us to deprecate `.` in version 2, whereas without this move we are stuck with `.` for ever.

Worth making this change and keeping a log of things that should be removed in version 2?

**Review by byrichardpowell — APPROVED** [review]
